### PR TITLE
use ffmpeg_qsv to decode and process video

### PR DESCRIPTION
**Pull request by FangleixLiu**

### CODE REVIEW

File: src/caffe/layers/video_data_layer.cpp [hunk 1]
```diff
@@ -37,7 +37,9 @@ void VideoDataLayer<Dtype>::DataLayerSetUp(
   video_type_ = video_data_param.video_type();
   skip_frames_ = video_data_param.skip_frames();
   CHECK_GE(skip_frames_, 0);
-
+  int t_w = this->transform_param_.resize_param().width();
+  int t_h = this->transform_param_.resize_param().height();
```
Comment (gongzg) [inline src/caffe/layers/video_data_layer.cpp]: For the two major object detection layers: SSD/YOLO2. We need to support the following transformation options:
1. Color format is RGB or BGR
2. FIT_LARGE_SIZE_AND_PAD mode resize. In this mode, we need to let the user to specify the padding value.
3. Resize_mode_WARP mode.
4. value scale and subtract specified mean value. For example, the output value need to subtract some specified mean value then multiply 1/255 or 1/127.5. The purpose is to normalize the input value to  [0,1] or (-1, 1).

File: src/caffe/layers/video_data_layer.cpp [hunk 2]
```diff
@@ -37,7 +37,9 @@ void VideoDataLayer<Dtype>::DataLayerSetUp(
...
+  //printf("t_w[%d] t_h[%d]\n", t_w, t_h);
```
Comment (gongzg) [inline src/caffe/layers/video_data_layer.cpp]: Please remove the above line of code.

File: src/caffe/layers/video_data_layer.cpp [hunk 3]
```diff
@@ -49,15 +51,28 @@ void VideoDataLayer<Dtype>::DataLayerSetUp(
   } else if (video_type_ == VideoDataParameter_VideoType_VIDEO) {
     CHECK(video_data_param.has_video_file()) << "Must provide video file!";
     const string& video_file = video_data_param.video_file();
+#ifndef USE_FFMPEG_QSV
     if (!cap_.open(video_file)) {
       LOG(FATAL) << "Failed to open video: " << video_file;
     }
     total_frames_ = cap_.get(CV_CAP_PROP_FRAME_COUNT);
+
```
Comment (gongzg) [inline src/caffe/layers/video_data_layer.cpp]: remove the blank line.

File: src/caffe/layers/video_data_layer.cpp [hunk 4]
```diff
@@ -49,15 +51,28 @@ void VideoDataLayer<Dtype>::DataLayerSetUp(
...
     processed_frames_ = 0;
     // Read image to infer shape.
     cap_ >> cv_img;
+
```
Comment (gongzg) [inline src/caffe/layers/video_data_layer.cpp]: ditto.

**Comment by gongzg** [comment]
@FangleixLiu another comment is how about to add another feature to generate a FP16 format output directly?

File: src/caffe/layers/video_data_layer.cpp (hunk 4 above)
Comment (FangleixLiu) [reply src/caffe/layers/video_data_layer.cpp]: delete blank line

File: src/caffe/layers/video_data_layer.cpp (hunk 3 above)
Comment (FangleixLiu) [reply src/caffe/layers/video_data_layer.cpp]: delete blank line

File: src/caffe/layers/video_data_layer.cpp (hunk 2 above)
Comment (FangleixLiu) [reply src/caffe/layers/video_data_layer.cpp]: remove the printf

**Comment by gongzg** [comment]
@FangleixLiu do you plan to add those preprocessing features in another PR?

**Comment by FangleixLiu** [comment]
about FP16 format output, now the MSS does not support FP16 output, maybe use opencl to do it

**Comment by FangleixLiu** [comment]
I look the source, in Transform function will do resize by opencl, so remove the resize by MSS

**Comment by gongzg** [comment]
Please check the function AspectKeepingResizeAndPad(), clCaffe haven't implemented OCL version resize. It is using the cv::resize and cv::copyMakeBorder to do the resizing and padding. But anyway, this patch itself is good enough. Let's leave the acceleration of resizing to the other patches. Thanks for your contribution.

**Pull request closed without merging**